### PR TITLE
[transformer] Fuse grad accumulation with wgrad

### PR DESCRIPTION
**Pull request by crcrpar**
In ColumnParallelLInear, we can fuse gradient accumulation into weight grad calc.

### CODE REVIEW

File: apex/transformer/tensor_parallel/layers.py [hunk 1]
```diff
@@ -203,15 +223,66 @@ def forward(self, input_):
...
+        grad_input = grad_output.matmul(weight)
+        # Convert the tensor shapes to 2D for execution compatibility
+        grad_output = grad_output.view(grad_output.shape[0] * grad_output.shape[1], grad_output.shape[2])
+        input = input.view(input.shape[0] * input.shape[1], input.shape[2])
+        if ctx.async_grad_allreduce:
+            # Asynchronous all-reduce
+            handle = torch.distributed.all_reduce(grad_input, group=get_tensor_model_parallel_group(), async_op=True)
+            # Delay the start of weight gradient computation shortly (3us) to have
+            # all-reduce scheduled first and have GPU resources allocated
+            _ = torch.empty(1, device=grad_output.device) + 1
```
Comment (eqy) [inline apex/transformer/tensor_parallel/layers.py:252]: Is this guaranteed to cause the `all_reduce` to happen first?

File: tests/L0/run_transformer/run_layers_test.py [hunk 2]
```diff
@@ -220,28 +220,36 @@ def test_column_parallel_linear(tensor_model_parallel_size):
     output_size_coeff = 17
     output_size = output_size_coeff * tensor_model_parallel_size
     batch_size = 7
+    hidden_size = 9
 
     # Network
-    identity_layer = IdentityLayer2D(batch_size, input_size).cuda()
+    identity_layer = IdentityLayer3D(batch_size, hidden_size, input_size).cuda()
     linear_layer = layers.ColumnParallelLinear(
```
Comment (eqy) [inline tests/L0/run_transformer/run_layers_test.py:227]: is `gradient_accumulation_fusion` set to `True` here?

Comment (crcrpar) [reply tests/L0/run_transformer/run_layers_test.py:227]: oh, I forgot to commit that change, appreciate it.

File: apex/transformer/tensor_parallel/layers.py (hunk 1 above)
Comment (crcrpar) [reply apex/transformer/tensor_parallel/layers.py:252]: experimentally yes.

File: csrc/megatron/fused_weight_gradient_dense_cuda.cu [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    } else {
+        d_output_2d = d_output;
+    }
+
+    const int hidden_dim = input_2d.size(0);
+    const int in_dim = input_2d.size(1);
+    const int out_dim = d_weight.size(0);
+
+    DISPATCH_FLOAT_HALF_AND_BFLOAT(input_2d.scalar_type(), 0, "wgrad_gemm_accum_fp32",
+        wgrad_gemm_accum_fp32_cuda<scalar_t_0>(
```
Comment (eqy) [inline csrc/megatron/fused_weight_gradient_dense_cuda.cu:187]: Haven't seen `scalar_t_0` before, what's the difference from regular `scalar_t`?

File: setup.py [hunk 4]
```diff
@@ -297,6 +297,41 @@ def append_nvcc_threads(nvcc_extra_args):
         )
     )
 
+    # Check, if CUDA11 is installed for compute capability 8.0
+    cc_flag = []
+    _, bare_metal_major, _ = get_cuda_bare_metal_version(CUDA_HOME)
+    if int(bare_metal_major) >= 11:
```
Comment (eqy) [inline setup.py:303]: Would we also want to include `sm_86`?

**Review by eqy — APPROVED** [review]
LGTM

File: csrc/megatron/fused_weight_gradient_dense_cuda.cu (hunk 3 above)
Comment (crcrpar) [reply csrc/megatron/fused_weight_gradient_dense_cuda.cu:187]: Macros in "type_shim.h" take `level` argument (in this case, 0) and allow us to nest dispatchers, e.g.
https://github.com/NVIDIA/apex/blob/082f999a6e18a3d02306e27482cc7486dab71a50/csrc/multi_tensor_lamb_stage_1.cu#L131-L146

File: setup.py (hunk 4 above)
Comment (crcrpar) [reply setup.py:303]: done in 5100286a18afb58e10fccc491642f9e4990ad75b

File: tests/L0/run_transformer/run_layers_test.py [hunk 5]
```diff
@@ -629,6 +642,7 @@ def test_parallel_transformer_layer(tensor_model_parallel_size):
 
     print_separator('test column-parallel linear')
     tensor_model_parallel_size = 1
+    test_column_parallel_linear(tensor_model_parallel_size)
```
Comment (crcrpar) [inline tests/L0/run_transformer/run_layers_test.py:645]:
```suggestion

```

**Pull request merged**